velox/functions/sparksql/ToJson.cpp

```diff
@@ -0,0 +1,540 @@
+#include <type/Type.h>
+#include <utility>
+#include <vector/BaseVector.h>
+#include <vector/ComplexVector.h>
+#include <vector/SelectivityVector.h>
+#include <vector/SimpleVector.h>
+#include <vector/TypeAliases.h>
+#include <vector/VectorEncoding.h>
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+template <typename T>
+void generateJsonTyped(
+    const SimpleVector<T>& input,
+    int row,
+    std::string& result,
+    const TypePtr& type) {
+  auto value = input.valueAt(row);
+
+  if constexpr (std::is_same_v<T, StringView>) {
+    size_t resultSize = escapedStringSize(value.data(), value.size());
+    result.resize(resultSize + 2);
+    result.data()[0] = '"';
+    escapeString(value.data(), value.size(), result.data() + 1);
+    result.data()[resultSize + 1] = '"';
+  } else if constexpr (std::is_same_v<T, UnknownValue>) {
+    VELOX_FAIL(
+        "Convert UNKNOWN to JSON: Vextors of UNKNOWN type should not contain non-null rows");
+  } else {
+    if constexpr (std::is_same_v<T, bool>) {
+      result.append(value ? "true" : "false");
+    } else if constexpr (
+        std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      if (FOLLY_UNLIKELY(std::isinf(value) || std::isnan(value))) {
+        result.append(fmt::format(
+            "\"{}\"",
+            util::Converter<TypeKind::VARCHAR>::tryCast(value).value()));
+      } else {
+        result.append(
+            util::Converter<TypeKind::VARCHAR>::tryCast(value).value());
+      }
+    } else if constexpr (std::is_same_v<T, Timestamp>) {
+      std::string stringValue = std::to_string(value);
+      result.reserve(stringValue.size() + 2);
+      result.append("\"");
+      result.append(stringValue);
+      result.append("\"");
+    } else if (type->isDate()) {
+      std::string stringValue = DATE()->toString(value);
+      result.reserve(stringValue.size() + 2);
+      result.append("\"");
+      result.append(stringValue);
+      result.append("\"");
+    } else if (type->isDecimal()) {
+      result.append(DecimalUtil::toString(value, type));
+    } else {
+      folly::toAppend<std::string, T>(value, &result);
+    }
+  }
+}
+
+// Convert primitive-type input vectors to Json string.
+template <
+    TypeKind kind,
+    typename std::enable_if_t<TypeTraits<kind>::isPrimitiveType, int> = 0>
+void toJson(
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  // input is guaranteed to be in flat or constant encodings when passed in.
+  auto inputVector = input.as<SimpleVector<T>>();
+
+  std::string result;
+  context.applyToSelectedNoThrow(rows, [&](auto row) {
+    if (inputVector->isNullAt(row)) {
+      flatResult.set(row, "null");
+    } else {
+      result.clear();
+      generateJsonTyped<T>(*inputVector, row, result, inputVector->type());
+
+      flatResult.set(row, StringView(result));
```

Comment:
For the to-string functions, usually we can optimize by estimating total size, allocating the data buffer once and writing to the data buffer directly. This helps avoid generating intermediate strings and improve the performance. Some example:
https://github.com/facebookincubator/velox/blob/main/velox/expression/CastExpr-inl.h#L609-L624